clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -189,21 +189,29 @@ protected abstract void onJoinComplete(int generation,
      * Block until the coordinator for this group is known and is ready to receive requests.
      */
     public synchronized void ensureCoordinatorReady() {
-        while (coordinatorUnknown()) {
+        ensureCoordinatorReady(Long.MAX_VALUE);
+    }
+
+    public synchronized void ensureCoordinatorReady(long timeoutMs) {
+        long remainingMs = timeoutMs;
+        long startTimeMs = time.milliseconds();
+        while (coordinatorUnknown() && remainingMs > 0) {
             RequestFuture<Void> future = lookupCoordinator();
-            client.poll(future);
+            client.poll(future, remainingMs);
 
             if (future.failed()) {
-                if (future.isRetriable())
-                    client.awaitMetadataUpdate();
-                else
+                if (future.isRetriable()) {
+                    remainingMs = timeoutMs == Long.MAX_VALUE ? Long.MAX_VALUE : time.milliseconds() - startTimeMs;
```

Comment:
Shouldn't this be `timeoutMs - (time.milliseconds() - startTimeMs)`? Also, it's not too big of a deal, but the checks for `Long.MAX_VALUE` seem like overkill.